pkgs/stdenv/generic/make-derivation.nix

```diff
@@ -439,6 +446,27 @@ else let
         "/bin/sh"
       ];
       __propagatedImpureHostDeps = computedPropagatedImpureHostDeps ++ __propagatedImpureHostDeps;
+    } //
+    # If we use derivations directly here, they end up as build-time dependencies.
+    # This is especially problematic in the case of disallowed*, since the disallowed
+    # derivations will be built by nix as build-time dependencies, while those
+    # derivations might take a very long time to build, or might not even build
+    # successfully on the platform used.
+    # However, in order to scan for disallowed references in the final output,
+    # knowing the out path is sufficient, and the out path can be calculated from
+    # the derivation without needing to build it, saving time and resources.
```

Comment:
how does this interact with CA derivations

Reply:
That should probably work (although a bit fragile) because `drv.outPath` will evaluate to a placeholder that will be rewritten just before the build if `drv` is part of the derivation closure. Which means that
1. If `drv` is part of the closure (because it's refered to somewhere else), then it will do the right thing (the `unsafeDiscardStringContext` will morally be a no-op)
2. Otherwise, the placeholder won't be rewritten, but it's fine since because the derivation isn't part of the build-time closure we know that it can't be in the runtime closure either